File: demo-deno-tf/test/calculator.test.ts
import { expect } from "jsr:@std/expect";
import { assertThrows } from "jsr:@std/assert";
import { add, divide, multiply, subtract } from "../src/calculator.ts";

Deno.test("add: should add positive numbers", () => {
  expect(add(5, 3)).toBe(8);
});

Deno.test("add: should add negative numbers", () => {
  expect(add(-2, -3)).toBe(-5);
});

Deno.test("add: should add mixed numbers", () => {
  expect(add(-5, 10)).toBe(5);
});

Deno.test("add: should add zeros", () => {
  expect(add(0, 0)).toBe(0);
});

Deno.test("subtract: should subtract positive numbers", () => {
  expect(subtract(10, 5)).toBe(5);
});

Deno.test("subtract: should subtract negative numbers", () => {
  expect(subtract(-5, -3)).toBe(-2);
});

Deno.test("subtract: should subtract mixed numbers", () => {
  expect(subtract(5, 10)).toBe(-5);
});

Deno.test("subtract: should subtract zeros", () => {
  expect(subtract(0, 0)).toBe(0);
});

Deno.test("multiply: should multiply positive numbers", () => {
  expect(multiply(5, 3)).toBe(15);
});

Deno.test("multiply: should multiply negative numbers", () => {
  expect(multiply(-2, -3)).toBe(6);
});

Deno.test("multiply: should multiply mixed numbers", () => {
  expect(multiply(-5, 10)).toBe(-50);
});

Deno.test("multiply: should multiply by zero", () => {
  expect(multiply(5, 0)).toBe(0);
});

Deno.test("divide: should divide positive numbers", () => {
  expect(divide(10, 2)).toBe(5);
});

Deno.test("divide: should divide negative numbers", () => {
  expect(divide(-6, -3)).toBe(2);
});

Deno.test("divide: should divide mixed numbers", () => {
  expect(divide(-10, 5)).toBe(-2);
});

Deno.test("divide: should throw when dividing by zero", () => {
  assertThrows(() => divide(5, 0), Error, "Division by zero");
});

File: demo-golang-tf/test/calculator_test.go
package calculator_test

import (
	"testing"

	"github.com/demo-tf/calculator/calculator"
)

func TestAdd(t *testing.T) {
	// Test cases
	tests := []struct {
		name     string
		a        int
		b        int
		expected int
	}{
		{"Positive numbers", 5, 3, 8},
		{"Negative numbers", -2, -3, -5},
		{"Mixed numbers", -5, 10, 5},
		{"Zeros", 0, 0, 0},
	}

	// Run test cases
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := calculator.Add(tt.a, tt.b)
			if result != tt.expected {
				t.Errorf("Addition result should match expected value: expected %d, got %d", tt.expected, result)
			}
		})
	}
}

func TestSubtract(t *testing.T) {
	tests := []struct {
		name     string
		a        int
		b        int
		expected int
	}{
		{"Positive numbers", 10, 5, 5},
		{"Negative numbers", -5, -3, -2},
		{"Mixed numbers", 5, 10, -5},
		{"Zeros", 0, 0, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := calculator.Subtract(tt.a, tt.b)
			if result != tt.expected {
				t.Errorf("Subtraction result should match expected value: expected %d, got %d", tt.expected, result)
			}
		})
	}
}

func TestMultiply(t *testing.T) {
	tests := []struct {
		name     string
		a        int
		b        int
		expected int
	}{
		{"Positive numbers", 5, 3, 15},
		{"Negative numbers", -2, -3, 6},
		{"Mixed numbers", -5, 10, -50},
		{"Zero multiplication", 5, 0, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := calculator.Multiply(tt.a, tt.b)
			if result != tt.expected {
				t.Errorf("Multiplication result should match expected value: expected %d, got %d", tt.expected, result)
			}
		})
	}
}

func TestDivide(t *testing.T) {
	// Test cases for successful division
	tests := []struct {
		name     string
		a        int
		b        int
		expected int
		hasError bool
	}{
		{"Positive numbers", 10, 2, 5, false},
		{"Negative numbers", -6, -3, 2, false},
		{"Mixed numbers", -10, 5, -2, false},
		{"Division by zero", 5, 0, 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := calculator.Divide(tt.a, tt.b)
			
			if tt.hasError {
				if err == nil {
					t.Error("Expected an error for division by zero")
				}
			} else {
				if err != nil {
					t.Errorf("Did not expect an error for valid division: %v", err)
				}
				if result != tt.expected {
					t.Errorf("Division result should match expected value: expected %d, got %d", tt.expected, result)
				}
			}
		})
	}
} 